### PR TITLE
Feat/add circuit inversion

### LINKED ISSUE #39: Add function that inverts quantum circuits

### PR DESCRIPTION
**Pull request by npoirie1**
Adds functions for generating the inverse of circuits and gates. The Gate struct was also replaced by a Gate abstract type. This means that all gate types have their own struct which are descendants of the Gate abstract type. This improves performance since an Operator is no longer generated every time a gate is constructed. The use of dispatching also improves the extensibility of the get_inverse function.

Fixes #39

### CODE REVIEW

File: docs/src/library.md [hunk 1]
```diff
@@ -58,6 +64,7 @@ Ket
 Bra
 Operator
 Base.adjoint
+ishermitian
```
Comment (alireza-najafi) [inline docs/src/library.md:67]: please change the name to `is_hermitian`.

File: src/Snowflake.jl [hunk 2]
```diff
@@ -41,6 +41,7 @@ export
     create,
     destroy,
     number_op,
+    ishermitian,
```
Comment (alireza-najafi) [inline src/Snowflake.jl:44]: change to `is_hermitian`.

File: src/core/qobj.jl [hunk 3]
```diff
@@ -112,6 +112,38 @@ Compute the adjoint (a.k.a. conjugate transpose) of a Ket, a Bra, or an Operator
 Base.adjoint(x::Ket) = Bra(x)
 Base.adjoint(x::Bra) = Ket(adjoint(x.data))
 Base.adjoint(A::Operator) = Operator(adjoint(A.data))
+
+"""
+    ishermitian(A::Operator)
```
Comment (alireza-najafi) [inline src/core/qobj.jl:117]: `is_hermitian`

File: src/core/qobj.jl [hunk 4]
```diff
@@ -112,6 +112,38 @@ Compute the adjoint (a.k.a. conjugate transpose) of a Ket, a Bra, or an Operator
...
+0.0 + 0.0im    1.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im
+
+
+julia> ishermitian(P)
+false
+
+```
+"""
+ishermitian(A::Operator) = LinearAlgebra.ishermitian(A.data)
```
Comment (alireza-najafi) [inline src/core/qobj.jl:145]: change to `is_hermitian`

File: test/qobj.jl [hunk 5]
```diff
@@ -138,4 +138,9 @@ end
                                 0 0 1])
     @test get_num_bodies(kron(qutrit_operator, qutrit_operator),
         hilbert_space_size_per_qutrit) == 2
+end
+
+@testset "ishermitian" begin
+    @test ishermitian(sigma_y())
```
Comment (alireza-najafi) [inline test/qobj.jl:144]: change to `is_hermitian`.

**Review by alireza-najafi — CHANGES_REQUESTED** [review]
This is great!. I suggest we change `ishermitian` to `is_hermitian` to be more consistent with the coding style as well as making it more readable.

**Comment by npoirie1** [comment]
The ishermitian function has been renamed to is_hermitian.

**Review by alireza-najafi — APPROVED** [review]
Thanks! Great Job.

**Pull request merged**